### PR TITLE
Cache services.ClusterConfig in AuthServer

### PR DESCRIPTION
**Pull request by russjones**
**Purpose**

As part of #1327, the proxy has the requirement that it should be configurable at runtime to either record at the proxy or record at the node. To enable this, we we cache `services.ClusterConfig` in the Auth Server and then use this cached value when authorizing requests. This allows us to change which permissions are returned for the proxy at runtime.

**Implementation**

* Cache `services.ClusterConfig` in the Auth Server.
* Pass a getter for `services.ClusterConfig` along in the context for every request to the Auth Server.
* Check where sessions are being recorded in `permissions.go` and return a `services.RoleSet` with `RW` access to `services.HostCert` if the sessions are being recorded at the proxy.

**Related Issues**

Fixes https://github.com/gravitational/teleport/issues/1441

### CODE REVIEW

File: lib/auth/permissions.go [hunk 1]
```diff
@@ -181,6 +181,38 @@ func GetCheckerForBuiltinRole(role teleport.Role) (services.AccessChecker, error
 				},
 			})
 	case teleport.RoleProxy:
+		if clusterConfig.GetSessionRecording() == services.RecordAtProxy {
```
Comment (kontsevoy) [inline lib/auth/permissions.go]: a comment here would be nice. what's going on? seems like a special hard-coded role is getting defined.

File: lib/defaults/defaults.go [hunk 2]
```diff
@@ -188,6 +188,10 @@ var (
 
 	// ReportingPeriod is a period for reports in logs
 	ReportingPeriod = 5 * time.Minute
+
+	// UpdateClusterConfigPeriod is how often the cached cluster config is
+	// updated within the Auth Server.
+	UpdateClusterConfigPeriod = 10 * time.Second
```
Comment (kontsevoy) [inline lib/defaults/defaults.go]: if we're polling, can we re-use one of existing constants? for example our nodes ping also (similar concept - also related to cluster state), so it would be easier to keep the same TTL in mind if you're doing support/troubleshooting?

**Review by kontsevoy — APPROVED** [review]
Approved, but I added some minor suggestions to consider.

File: lib/auth/tun.go [hunk 3]
```diff
@@ -358,7 +358,10 @@ func (s *AuthTunnel) onAPIConnection(sconn *ssh.ServerConn, sshChan ssh.Channel,
 			log.Error(err.Error())
 			return
 		}
-		user = teleport.BuiltinRole{Role: systemRole}
+		user = BuiltinRole{
+			ClusterConfigGetter: s.authServer.getCachedClusterConfig,
```
Comment (klizhentas) [inline lib/auth/tun.go]: can you rename to `GetClusterConfig` instead of `Getter`

File: lib/auth/permissions.go (hunk 1 above)
Comment (klizhentas) [inline lib/auth/permissions.go]: the previous version was just adding a couple of rules, why did you change it to the whole new role?

**Comment by klizhentas** [comment]
overall looks good, just a couple of nitpicks

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Review by klizhentas — APPROVED** [review]

**Comment by russjones** [comment]
@gravitational-jenkins retest this please
